job-info, kvs-watch: support guest disconnect & credential checks

chu11 opened the pull request:
Since the fixes for #3526, #3528, #3529 are similar and all related, I've put them all in this PR, although we could split them out.

1) Add a new function `flux_msg_match_cred()` to handle the common case check of ensuring a message came from an instance owner or the userid of the sender matches another message's userid.

2) Clean up some cancel/disconnect code that manually matched message senders and instead use `flux_msg_match_route_first()`

3) Add a call to `flux_msg_match_cred()` to also check for a guest user credentials.

I didn't add any tests as I'm confident that existing tests already cover this (since its cancel paths) or we've generally never covered (disconnect).

garlick commented on src/common/libflux/message.c:
```diff
@@ -1675,7 +1675,25 @@ bool flux_msg_match_route_first (const flux_msg_t *msg1, const flux_msg_t *msg2)
     return true;
 }
 
+bool flux_msg_match_cred (const flux_msg_t *msg1, const flux_msg_t *msg2)
```
Would `flux_msg_authrorize()` work here?

chu11 replied:
oh, i had completely missed that you wrote `flux_msg_authorize()`, yeah, that would simplify the function.

chu11 replied:
I didn't know about `flux_msg_authorize()` b/c I saw in `libflux/disconnect.c` that you hand wrote an authorize function.

```
static bool authorize (struct flux_msg_cred cred, const flux_msg_t *msg)
{
    uint32_t userid;

    if ((cred.rolemask & FLUX_ROLE_OWNER))
        return true;
    if ((cred.rolemask & FLUX_ROLE_USER)
        && flux_msg_get_userid (msg, &userid) == 0
        && cred.userid == userid)
        return true;
    return false;
}
```

so perhaps we can clean up some code in `disconnect.c` as well.

garlick reviewed (commented):
Would anything in `src/common/libflux/disconnect.[ch]` help here or is that too much surgery?

chu11 commented:
> Would anything in src/common/libflux/disconnect.[ch] help here or is that too much surgery?

Yeah, I thought it was just a little bit too much surgery.  If there was a way to do a callback w/ the msglist on messages that matched, that might have worked out.  But thought that to be too much effort to add to the disconnect lib.

chu11 commented:
re-worked how I did `flux_msg_match_cred()`.  Basically the right combo of helper functions didn't exist, so I did a bit of refactoring to create some new internal functions in `message.c`.  Basically, I wanted to avoid a call to `flux_msg_get_userid()` if I didn't need to call it.  Then cleaned up some code in `disconnect.c`.  Fixed up a typo I found along the way too.

garlick reviewed (commented):
> Basically, I wanted to avoid a call to flux_msg_get_userid() if I didn't need to call it. Then cleaned up some code in disconnect.c

I think what ended up happening is now the loop in disconnect.c has to call `flux_msg_get_cred()` on the disconnect/cancel request for each element in the msglist, so it seems this will be a wash performance wise.  And anyway, I _think_ accessing anything in the message "proto block" should be pretty fast (certainly will be after #3617, if those become struct elements).

If the performance argument is moot, and if you agree with my comment that this is disconnect/cancel specific, does it make more sense to move the helper function to `disconnect.c`?  Then at least the disconnect/cancel related message stuff is all in one place from a security audit perspective.

chu11 commented:
> I think what ended up happening is now the loop in disconnect.c has to call flux_msg_get_cred() on the disconnect/cancel request for each element in the msglist, so it seems this will be a wash performance wise. And anyway, I think accessing anything in the message "proto block" should be pretty fast (certainly will be after #3617, if those become struct elements).

Thinking about this a tad, perhaps given how the convenience function ended up being used, there should be a `flux_match_cred()` helper function and then an additional `flux_match_route_first_and_cred()` function (name TBD).  That way we can avoid the excess `flux_msg_get_cred()` on each message.

garlick commented:
I can't tell if you disagree with my suggestions or if I failed to be clear :-).  In summary, I meant:
* the overhead of `flux_msg_get_cred()` and `flux_msg_get_userid()` is probably (or soon will be) negligible
* message methods that take two messages and check if first message is allowed to do something to second message is likely only applicable to disconnect and cancel
* lets expand disconnect.c to meet the need here (even if it's just adding a single helper there) not message.c

chu11 commented:
@garlick oh yeah, generally speaking agree.  Your comment about calling `flux_msg_get_cred()` just made me wonder if a helper function to check `route_first` and `authorize` would be better.  Haven't decided yet, will see as I implement :-)

chu11 commented:
re-pushed, now trying a slightly different way.

- wrote a new function `flux_msg_match_route_cred()` (not the biggest fan of the function name, suggestions welcome).  It does `flux_msg_match_route_first()` and then `flux_msg_cred_authorize()` and put it in the disconnect code.  Since we consider `flux_msg_get_cred()` and `flux_msg_get_userid()` to be negligible, they are retrieved before every authenticate check.

- now, all the places where there was previously a `flux_msg_match_route_first()` and `flux_msg_match_cred()`/`authenticate()` call, is replaced with a single call to `flux_msg_match_route_cred()`.

garlick commented:
Ah better!  Would it make sense to have a `flux_disconnect_match()` and `flux_cancel_match()`?  The cancel one could call `flux_disconnect_match()`, then on success go on to parse the matchtag from the payload of `msg1` and match it to `msg2`?

chu11 commented:
> Ah better! Would it make sense to have a flux_disconnect_match() and flux_cancel_match()? The cancel one could call flux_disconnect_match(), then on success go on to parse the matchtag from the payload of msg1 and match it to msg2?

Ahh that's a good idea!  I couldn't come up with a good name b/c the matching was different between cancel & disconnect.

chu11 commented:
re-pushed, now there are 2 helper functions `flux_disconnect_match()` and `flux_cancel_match()`, with cleanups in `job-info` and `kvs-watch`.  One of the nice fallouts is we don't have to parse `matchtag` and manually check it anymore.

garlick commented on src/modules/job-info/watch.c:
```diff
@@ -468,14 +466,12 @@ void watch_cancel_cb (flux_t *h, flux_msg_handler_t *mh,
-
-    if (flux_request_unpack (msg, NULL, "{s:i}", "matchtag", &matchtag) < 0) {
-        flux_log_error (h, "%s: flux_request_unpack", __FUNCTION__);
+    if (flux_request_decode (msg, NULL, NULL) < 0) {
```
Checking that the cancel request is a request is not super useful, since that would have already been checked before dispatching the message to this callback, and we can't respond with EPROTO to it since it is a noresponse.

chu11 replied:
I wasn't sure if it was necessary as well, but I noticed a `flux_request_decode()` in the disconnect callback.  Should it be removed there as well?  The pattern is in multiple modules (minimally `job-info`, `kvs`, and `kvs-watch`).

garlick commented on src/common/libflux/disconnect.c:
```diff
@@ -16,32 +16,31 @@
+    if (flux_msg_get_userid (msg2, &userid) < 0)
+        return false;
+
+    return !flux_msg_cred_authorize (cred, userid) ? true : false;
```
IMHO this would be a bit simpler if it was just:
```c
if (flux_msg_cred_authorize (cred, userid) < 0)
    return false;
return true;
```

garlick commented on src/common/libflux/disconnect.h:
```diff
@@ -17,11 +17,23 @@
 extern "C" {
 #endif
 
+/* Convenience function for ensuring messages match for disconnect.
+ * Effectively calls flux_msg_match_route_first() and
+ * flux_msg_cred_authorize().
+ */
```
Might be clearer to say "Return true if disconnect request `msg1` came from same sender as `msg2' and has appropraite authorization".

garlick commented on src/common/libflux/disconnect.h:
```diff
@@ -17,11 +17,23 @@
  */
 int flux_msglist_disconnect (struct flux_msglist *l, const flux_msg_t *msg);
 
+/* Convenience function for ensuring messages match for cancel.
+ * Effectively calls flux_disconnect_match() and flux_msg_get_matchtag() to
+ * compare matchtags.
+ */
```
How about "Return true if cancel request `msg1` came from same sender as `msg2`, has appropriate authorization, and references the matchtag in `msg2`.

garlick reviewed (approved):
Looks good!  Thanks for the iteration.  Just some minor comments.

garlick commented:
Probably superfluous, as long as the message handler is registered for a
specific topic string and requests only (not a mask of types).

Jim

On Wed, May 5, 2021 at 8:56 AM Al Chu ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/modules/job-info/watch.c
> <https://github.com/flux-framework/flux-core/pull/3627#discussion_r626694814>
> :
>
> > @@ -468,14 +466,12 @@ void watch_cancel_cb (flux_t *h, flux_msg_handler_t *mh,
>                        const flux_msg_t *msg, void *arg)
>  {
>      struct info_ctx *ctx = arg;
> -    uint32_t matchtag;
> -
> -    if (flux_request_unpack (msg, NULL, "{s:i}", "matchtag", &matchtag) < 0) {
> -        flux_log_error (h, "%s: flux_request_unpack", __FUNCTION__);
> +    if (flux_request_decode (msg, NULL, NULL) < 0) {
>
> I wasn't sure if it was necessary as well, but I noticed a
> flux_request_decode() in the disconnect callback. Should it be removed
> there as well? The pattern is in multiple modules (minimally job-info, kvs,
> and kvs-watch).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/3627#discussion_r626694814>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABJPW47JBJ3YN7MPNQDQMDTMFTCTANCNFSM436QTGTA>
> .
>

chu11 commented:
re-pushed

- fixed up things per @garlick review
- removed superfluous calls to `flux_request_decode()`, including in the `kvs` module
- fixed up some comments
- also used `flux_msg_match_route_first()` in the `kvs` module

garlick commented on src/modules/kvs-watch/kvs-watch.c:
```diff
@@ -703,43 +703,44 @@ static void watcher_respond_ns (struct ns_monitor *nsm)
         flux_log_error (nsm->ctx->h, "%s: zlist_dup", __FUNCTION__);
 }
 
```
In commit message for a3d9e2355, second "to" should be "do"?

> Problem: kvs-watch is open to guest users, but the cancellations
and disconnects of the watches to not verify that the request came from the
same guest user

garlick reviewed (approved):
Thanks!  I did find one small commit typo.  Set MWP when you're ready.

chu11 commented:
@garlick, thanks.  re-pushed with the commit message typo and setting MWP.

chu11 commented:
hmmm hit a flux-sched build error.   Can't find a log.  Assuming this isn't related to this PR b/c many prior runs were successful.

```
   PASS: t1009-recovery-multiqueue.t 1 - recovery: generate test jobspecs
  PASS: t1009-recovery-multiqueue.t 2 - load test resources
  PASS: t1009-recovery-multiqueue.t 3 - recovery: loading flux-sched modules with two queues
  PASS: t1009-recovery-multiqueue.t 4 - recovery: submit to occupy resources fully (rv1)
  PASS: t1009-recovery-multiqueue.t 5 - recovery: works when both modules restart (rv1)
  PASS: t1009-recovery-multiqueue.t 6 - recovery: a cancel leads to a job schedule (rv1)
  FAIL: t1009-recovery-multiqueue.t 7 - recovery: cancel all jobs (rv1_nosched)
  FAIL: t1009-recovery-multiqueue.t 8 - cleanup active jobs
  FAIL: t1009-recovery-multiqueue.t 9 - removing resource and qmanager modules
  ERROR: t1009-recovery-multiqueue.t - missing test plan
  tap-driver.sh: internal error getting exit status
  tap-driver.sh: fatal: I/O or internal error
  Makefile:830: recipe for target 't1009-recovery-multiqueue.log' failed
```
